eyg/src/home/mod.rs

```diff
@@ -11,20 +11,34 @@ pub trait Deliverable<S> {
 pub type Mail<S> = Vec<Box<dyn Deliverable<S>>>;
 
 pub trait Handler<M, S> {
-    fn handle(self, message: M) -> (Mail<S>, Self);
+    type Next;
+    fn handle(self, message: M) -> (Mail<S>, Self::Next);
 }
 
 // Expose that there are Actors or no useful insights
 // Home is the same as System or Environment or Rave or Workshop or Village,
 // Workshop is interesting as Home could become skill/capability
+// If this is a Home to these kind of workers if must be possible to insert them.
+// It's up to some other step to check that if state changes stuff still works
 pub trait Home<A, M>: Sized {
     type Worker: Handler<M, Self>;
     // I don't think this is very ergonomic Rust
     fn pop(self, address: &A) -> (Self::Worker, Self);
     fn put(self, address: A, worker: Self::Worker) -> Self;
 }
 
-impl<A, M, S> Deliverable<S> for Envelope<A, M> where S: Home<A, M> {
+// The specification of W here indicates this only applies for when a handler does not change the state of the worker
+// impl<A, M, S, W> Deliverable<S> for Envelope<A, M>
+//     where
+//         S: Home<A, M, Worker=W>,
+//         W: Handler<M, S, Next=S::Worker>
+// {
+impl<A, M, S> Deliverable<S> for Envelope<A, M>
+    where
+        S: Home<A, M>,
+        S::Worker: Handler<M, S, Next=S::Worker>
```

Comment:
I thought this would limit this trait being implemented only to cases when the Next type of the Handler is equal to the original handler.

It appears to have been implemented for all envelopes.

The test error I get when running `cargo test home_test` is
```
 type mismatch resolving `<home::tests::home_test::Counter as home::Handler<(), std::collections::HashMap<std::string::String, home::tests::home_test::Counter>>>::Next == home::tests::home_test::Counter`
  --> src/home/mod.rs:82:55                                                                                          
   |                                                                                                                 
82 |         let (_mail, environment) = Box::new(envelope).deliver(environment);                                     
   |                                                       ^^^^^^^ expected i32, found struct `home::tests::home_test::Counter`
   |                                                                                                                 
   = note: expected type `i32`                                                                                       
              found type `home::tests::home_test::Counter`                                                           
   = note: required because of the requirements on the impl of `home::Deliverable<std::collections::HashMap<std::string::String, home::tests::home_test::Counter>>` for `home::Envelope<std::string::String, ()>`
```

I was expecting an error more like Deliverable not implemented. :thinking: